### DISCUSSION TITLE
A way to detect if a page has text saved as geometry and needs OCR

### DISCUSSION THREAD

**Discussion by idoglanz** [pymupdf/PyMuPDF#3558] [Looking for help]
Hi! first of all this library is great and extremely useful!

Something I encountered and was wondering if you have an elegant solution for. 
A lot of our PDFs have text saved in weird fonts, and as such the text is saved as geometry and not text.
To solve this we just run OCR and use the text extracted.
My question is, can you think of an efficient way to detect if a page needs OCR? (i.e. it has text saved as geometry). One brute force way could be to run it anyhow, find some words and search for them in the original TextPage, but I'm trying to save on the initial OCR.

Thank you!

**Comment by JorjMcKie** [comment]
There is no precise formula to detect this ... only heuristics. So much for elegance 🤷‍♂️ 😒.
* Presumably you are already checking whether there is text on the page at all get_text() is empty or only contains white space.
* No annotations shown (`page.first_annot == None`).
* Image(s) covering the better part of the page.

If there is text on page given as drawings paths, we can at least enumerate a number of necessary - albeit not sufficient - conditions:
* very many `get_drawings()` paths, i.e. a long list, several thousand are typical
* many of the path rectangles should have dimensions that we typically also would associate with character bboxes. We can also assume that all those little paths are of type "f" (fill no stroke color). And maybe even most of them fill color black (0,0,0). For example see this page with some text in drawing format:
```python
mport pymupdf
doc = pymupdf.open()
page = doc.new_page()
page.insert_text((100,100), "Some text")
svg = page.get_svg_image()  # get SVG from page
svgdoc = pymupdf.open("svg",svg.encode())  # document from SVG source
pdfdata = svgdoc.convert_to_pdf()  # make a PDF from it again
svgpdf = pymupdf.open("pdf", pdfdata)
svgpage=svgpdf[0]
svgpaths=svgpage.get_drawings()  # ok: we should expect exactly 8 paths: 1 for each character
for p in svgpaths:
    print(p["rect"].width, p["rect"].height)  # width and height

    
6.302978515625 8.404083251953125  # capital "S"
5.2139892578125 6.1819610595703125  # "o", etc.
7.6120147705078125 5.9288482666015625
5.2030792236328125 6.1819610595703125
2.6400604248046875 7.601104736328125
5.2030792236328125 6.1819610595703125
5.01593017578125 5.76385498046875
2.6400604248046875 7.601097106933594
```

**Reply by JorjMcKie** [reply]
Ah, forgot a possible other condition:
Characters as paths should have `path["items"]` of types "l" and "c" exclusively - IAW there should never be "re" or "q" items.

**Reply by JorjMcKie** [reply]
As an aside, the above is a way to create nasty pages like this yourself ...

**Reply by idoglanz** [reply]
Thanks for the quick reply!

I think I'll checkout the path['items'] types and threshold that.

Regarding the paths, aside from colors, anything else you think I can use there? not sure I understood how to use the dimensions

**Reply by JorjMcKie** [reply]
> how to use the dimensions

The argument was that text mostly has typical font size of, say 10 or 11 ... in that range. Of course, font sizes 20 or 35 do also occur.
So drawn characters should have boundary box heights <= those font sizes, and the width should regularly be strictly less than height.
In my example, font size 11 was used. The maximum height (capital "S") we see is 8.4  - you see the relationship.

**Reply by idoglanz** [reply]
ahh yes that makes sense, thanks. I'll try and threshold the distribution of heights and widths (assuming too uniform means text I suppose)
